From bbe0f1be2503a01da63b2a5e6d0eb9aedd14a35f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Markus=20M=C3=BCtzel?= <markus.muetzel@gmx.de>
Date: Fri, 28 Aug 2026 13:57:59 +0200
Subject: [PATCH] Preserve order of MPI libraries also in complicated link
 trees.

The link library for `msmpi.dll` in MSYS2 is used to insert thin wrapper
functions for compatibility with gfortran. These wrapper functions depend
on symbols that are exported from the shim library `libmsmpifec.dll`
which is detected by FindMPI. That shim library must be kept *after*
`msmpi.dll` in link commands.
More complicated link trees can lead to a situation where the order of
these libraries gets reversed.
Avoid that by grouping the MPI libraries together.
---
 Modules/FindMPI.cmake | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/Modules/FindMPI.cmake b/Modules/FindMPI.cmake
index ebfd79fe3e..63483e77cb 100644
--- a/Modules/FindMPI.cmake
+++ b/Modules/FindMPI.cmake
@@ -1396,7 +1396,8 @@ macro(_MPI_create_imported_target LANG)
   # If the compiler links MPI implicitly, no libraries will be found as they're contained within
   # CMAKE_<LANG>_IMPLICIT_LINK_LIBRARIES already.
   if(MPI_${LANG}_LIBRARIES)
-    set_property(TARGET MPI::MPI_${LANG} PROPERTY INTERFACE_LINK_LIBRARIES "${MPI_${LANG}_LIBRARIES}")
+    # Keep libraries together preserving their order even in complicated link trees
+    set_property(TARGET MPI::MPI_${LANG} PROPERTY INTERFACE_LINK_LIBRARIES "$<LINK_GROUP:RESCAN,${MPI_${LANG}_LIBRARIES}>")
   endif()
   # Given the new design of FindMPI, INCLUDE_DIRS will always be located, even under implicit linking.
   set_property(TARGET MPI::MPI_${LANG} PROPERTY INTERFACE_INCLUDE_DIRECTORIES "${MPI_${LANG}_INCLUDE_DIRS}")
-- 
2.51.0.windows.2

